File: LL/src/Core/Domain/Models/Attributes/AttributeDefinition.cs
namespace Domain.Models.Attributes;

public enum AttributeUnit
{
    FlatPoints,
    PercentagePoints,
    Rating,
    HealthPerFiveSeconds,
    MultiplierInput
}

public enum AttributeStackingRule
{
    Additive,
    Multiplicative,
    Maximum,
    DerivedOnly
}

public enum AttributeCapKind
{
    None,
    Fixed,
    ContextDependent
}

public enum AttributeBenchmarkScenario
{
    PhysicalOffense,
    MagicalOffense,
    PeriodicOffense,
    PhysicalPressure,
    MagicalPressure,
    HealingSustain,
    StatusResilience,
    CrowdControlResilience,
    SummonOffense,
    MixedPressure,
    UnmitigatedPressure,
    BurstPressure,
    LongSustain
}

public sealed record AttributeDefinition(
    AttributeType AttributeType,
    string DisplayName,
    string Description,
    AttributeUnit Unit,
    AttributeStackingRule StackingRule,
    float MinimumValue,
    float? MaximumValue,
    AttributeCapKind CapKind,
    bool IsEquipmentEligible,
    bool IsContentFacing,
    int DisplayPrecision,
    string DisplaySuffix,
    AttributeType? ApprovedPrimarySource,
    IReadOnlyList<AttributeBenchmarkScenario> RelevantBenchmarkScenarios);

File: LL/src/Presentation/ll/src/app/shared/models/attribute-definition.ts
import { AttributeType } from './enums/attributeType';

export type AttributeUnit =
  | 'FlatPoints'
  | 'PercentagePoints'
  | 'Rating'
  | 'HealthPerFiveSeconds'
  | 'MultiplierInput';

export type AttributeStackingRule =
  | 'Additive'
  | 'Multiplicative'
  | 'Maximum'
  | 'DerivedOnly';

export type AttributeCapKind = 'None' | 'Fixed' | 'ContextDependent';

export interface AttributeDefinition {
  attributeType: AttributeType;
  displayName: string;
  description: string;
  unit: AttributeUnit;
  stackingRule: AttributeStackingRule;
  minimumValue: number;
  maximumValue: number | null;
  capKind: AttributeCapKind;
  isEquipmentEligible: boolean;
  isContentFacing: boolean;
  displayPrecision: number;
  displaySuffix: string;
  relevantBenchmarkScenarios: string[];
  equipmentDisplayName?: string;
  equipmentDescription?: string;
  equipmentUnit?: AttributeUnit;
  equipmentDisplayPrecision?: number;
  equipmentDisplaySuffix?: string;
}

let definitions = new Map<AttributeType, AttributeDefinition>();

export function setAttributeDefinitions(
  values: readonly AttributeDefinition[] | null | undefined,
): void {
  definitions = new Map(
    (values ?? []).map((definition) => [
      definition.attributeType,
      definition,
    ]),
  );
}

export function getAttributeDefinition(
  attribute?: string | null,
): AttributeDefinition | undefined {
  if (!attribute) return undefined;
  return definitions.get(attribute as AttributeType);
}
